---
title: '**Introduction to nf-core**'
output:
  html_document:
    toc: false
    toc_float: false
from: markdown+emoji
---

# What is nf-core?

![](../figs/1.2_nf-core.png){width=100%}

nf-core is a community effort to collect a curated set of analysis workflows built using Nextflow.

nf-core provides a standardized set of best practices, guidelines, and templates for building and sharing bioinformatics workflows. These workflows are designed to be modular, scalable, and portable, allowing researchers to easily adapt and run them on their own data and compute resources.

The nf-core community comprises a diverse group of bioinformaticians, developers, and researchers from around the world who collaborate on developing and maintaining a growing collection of high-quality workflows. These workflows cover a range of applications, including transcriptomics, proteomics, and metagenomics.

One of the key benefits of nf-core is that it promotes open development, testing, and peer review, ensuring that the workflows are robust, well-documented, and validated against real-world datasets. This helps to increase the reliability and reproducibility of bioinformatics analyses and ultimately enables researchers to accelerate their scientific discoveries.

nf-core is published in Nature Biotechnology! [Nat Biotechnol 38, 276–278 (2020). Nature Biotechnology](https://www.nature.com/articles/s41587-020-0439-x)

**Key Features of nf-core workflows**

- **Documentation**
  - nf-core workflows have extensive documentation covering installation, usage, and description of output files to ensure that you won't be left in the dark.
- **CI Testing**
  - Every time a change is made to the workflow code, nf-core workflows use continuous-integration testing to ensure that nothing has broken.
- **Stable Releases**
  - nf-core workflows use GitHub releases to tag stable versions of the code and software, making workflow runs totally reproducible.
- **Packaged software**
  - Pipeline dependencies are automatically downloaded and handled using Docker, Singularity, Conda, or other software management tools. There is no need for any software installations.
- **Portable and reproducible**
  - nf-core workflows follow best practices to ensure maximum portability and reproducibility. The large community makes the workflows exceptionally well-tested and easy to run.
- **Cloud-ready**
  - nf-core workflows are tested on AWS after every major release. You can even browse results live on the website and use outputs for your own benchmarking.

All nf-core workflows are open-source and community driven.

## Events

nf-core events are community-driven gatherings that provide a platform to discuss the latest developments in Nextflow and nf-core workflows. These events typically include a combination of community seminars, trainings, and hackathons, and are open to anyone who is interested in using and developing nf-core and its applications. Most events are held virtually, making them accessible to a global audience.

Upcoming events are listed on the [nf-core event page](https://nf-co.re/events) and announced on Slack and Twitter.

## Join the community!

There are several ways you can join the nf-core community. You are welcome to join any or all of these at any time!

::: {layout-ncol=5 layout-valign="center"}

[![](../figs/1.2_slack.png)](https://nf-co.re/join/slack)

[![](../figs/1.2_github.png)](https://github.com/nf-core)

[![](../figs/1.2_twitter.png)](https://twitter.com/nf_core)

[![](../figs/1.2_mastodon.png)](https://mstdn.science/@nf_core)

[![](../figs/1.2_youtube.png)](https://www.youtube.com/c/nf-core)

:::

> Questions about Nextflow that are not related to nf-core can be asked on the [Nextflow Slack](https://www.nextflow.io/blog/2022/nextflow-is-moving-to-slack.html).

## nf-core tooling

This tutorial will make use of the nf-core/tools, a set of helper tools for use with Nextflow workflows. These tools have been developed to provide a range of additional functionality using, developing and testing workflows.

nf-core/tools is written in Python and is available from the [Python Package Index (PyPI)](https://pypi.org/project/nf-core/).

```bash
pip install nf-core
```

Alternatively, nf-core/tools can also be installed from [Bioconda](https://anaconda.org/bioconda/nf-core):

```bash
conda install -c bioconda nf-core
```

You can nf-core `--version` options can be used to print your version of nf-core/tools:

```bash
nf-core --version
```

::: callout-tip

### **Challenge**{.unlisted}

Find out what version of nf-core/tools you have available using the nf-core `--version` option:

:::

::: {.callout-caution collapse="true"}

### Solution

Use the nf-core `--version` option to print the nf-core/tools version:

```bash
nf-core --version

![`nf-core list`](../figs/nf-core-version.svg)

```

:::

nf-core/tools has commands to help both users and developers. The tooling is for everyone. For users, the helper tools make it easier to run workflows. For developers, the helper tools make it easier to develop your workflows using best practices.

::: callout-tip

### **Challenge**{.unlisted}

Find out what other nf-core/tools options and commands are available using the `--help` option:

:::

::: {.callout-caution collapse="true"}

### Solution

Run the `--help` option to list the options, tools for users, and tools for developers

```bash
nf-core --help
```

:::

## Finding a nf-core workflow

There are currently 80 workflows (March 2023) available as part of nf-core. These workflows are at various stages of development with 49 released, 19 under development, and 12 archived. A full list of workflows, as well as their documentation, can be explored on the [nf-core website](https://nf-co.re/).

Unless you are actively developing workflow code, you do not need to clone the workflow code from GitHub and can use Nextflow’s built-in functionality to pull amd store the workflows for you. The Nextflow `pull` command to download and cache nf-core workflows from the [nf-core GitHub](https://github.com/nf-core/) repository:

```bash
nextflow pull nf-core/<PIPELINE>
```

Nextflow `run` will also automatically `pull` the workflow if it was not already available locally:

```bash
nextflow run nf-core/<PIPELINE>
```

Nextflow will `pull` the default git branch if a workflow version is not specified. This will be the master branch for nf-core workflows with a stable release. nf-core workflows use GitHub releases to tag stable versions of the code and software. You will always be able to run a previous version of a workflow once it is released using the `-revision` or `-r` flag.

::: callout-tip

### **Challenge**{.unlisted}

Try to `pull` the latest version of the `nf-core/rnaseq` workflow directly from GitHub using Nextflow:

:::

::: {.callout-caution collapse="true"}

### Solution

Use the Nextlfow `pull` command to download the `rnaseq` workflow from the `nf-core` GitHub repository.

```bash
nextflow pull nf-core/rnaseq
```

:::
