---
title: "**Designing your run command**"
output:
  html_document:
    toc: false
    toc_float: false
from: markdown+emoji
---

::: callout-tip
### Objectives{.unlisted}

- Use the nf-core documentation to select appropriate parameters for a run command 
- Write and run a nf-core rnaseq command on the command line 
- Explore workflow deployment and set up 
:::

### **Download the workflow code**

It can be very easy to lose track while working on the command line, especially when we're working with large datasets and complex commands as we do with bioinformatics workflows. To make sure we work reproducibly, we will be organising our workspace and using a local copy of the [nf-core/rnaseq](https://nf-co.re/rnaseq) workflow for all exercises. 

Start by creating a new directory for all of today's activities and move into it: 

```bash
mkdir ~/nfcore-workshop/session2 && cd $_
```

There are a number of ways to download a nf-core workflow to your machine. We recommend using git or the [nf-core tools](https://nf-co.re/tools/) utility. Today, we will download most recent version of the workflow from it's [GitHub repository](https://github.com/nf-core/rnaseq) with git. 

Clone the nf-core/rnaseq repository:
```bash
git clone https://github.com/nf-core/rnaseq.git
```

Check the workflow has been downloaded: 
```bash
ls -l rnaseq
```

Inside your `nf-core-rnaseq` workflow directory, you should see a number of files and subdirectories: 
```default
total 216
-rw-rw-r-- 1 ubuntu ubuntu 58889 Apr  4 03:40 CHANGELOG.md
-rw-rw-r-- 1 ubuntu ubuntu  9681 Apr  4 03:40 CITATIONS.md
-rw-rw-r-- 1 ubuntu ubuntu  9078 Apr  4 03:40 CODE_OF_CONDUCT.md
-rw-rw-r-- 1 ubuntu ubuntu  1096 Apr  4 03:40 LICENSE
-rw-rw-r-- 1 ubuntu ubuntu 10002 Apr  4 03:40 README.md
drwxrwxr-x 3 ubuntu ubuntu  4096 Apr  4 03:40 assets
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 bin
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 conf
drwxrwxr-x 3 ubuntu ubuntu  4096 Apr  4 03:40 docs
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 lib
-rwxrwxr-x 1 ubuntu ubuntu  2736 Apr  4 03:40 main.nf
drwxrwxr-x 4 ubuntu ubuntu  4096 Apr  4 03:40 modules
-rw-rw-r-- 1 ubuntu ubuntu 13970 Apr  4 03:40 modules.json
-rw-rw-r-- 1 ubuntu ubuntu 10903 Apr  4 03:40 nextflow.config
-rw-rw-r-- 1 ubuntu ubuntu 42576 Apr  4 03:40 nextflow_schema.json
-rw-rw-r-- 1 ubuntu ubuntu   359 Apr  4 03:40 pyproject.toml
drwxrwxr-x 4 ubuntu ubuntu  4096 Apr  4 03:40 subworkflows
-rw-rw-r-- 1 ubuntu ubuntu  1684 Apr  4 03:40 tower.yml
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 workflows
```

This may look very different to other Nextflow workflows you may have run or written before. The most important files and directories for us to understand, are:

|Feature                |Importance                                                         |
|-----------------------|-------------------------------------------------------------------|
|`conf/`                |Contains standard configuration files for various profiles that build on global settings set by `nextflow.config`|
|`main.nf`              |The executable Nextflow script that defines the structure and flow of the workflow. It calls `workflows/rnaseq.nf`|
|`modules/`             |Contains Nextflow processes used by the workflow. They are called by the `main.nf` file|
|`workflows/rnaseq.nf`  |All the modules, subworkflows, channels, workflow structure for running the rnaseq workflow |

::: {.callout-caution collapse="true"}
### **Alternate installation method**{.unlisted}

Using the nf-core tools utility, search for the rnaseq pipeline: 
```bash
nf-core list rnaseq
```

Then, download the correct pipeline: 
```bash
nf-core download nf-core/rnaseq
```

You will be prompted to select a version. Use your arrow keys to specify 3.10.1 and hit enter. 

> :no_entry: **BEWARE** :no_entry: this method will download a copy of the workflow with a different directory name and slightly different structure. If you choose to use this method, you will need to adjust commands in the upcoming lessons accordingly. 
:::

### **Build your run command**

All nf-core workflows are provided with sensible default settings that have broad applicability and comprensive documentation that explains all available parameters. What is 'sensible' varies dramatically between different experiments, computing environments, and datasets, so these settings might not suit your needs. Having asked ourselves those questions earlier, consider the most important design choices we have made for this workflow and structuring our run command:

* We **don't** need to run the pseudo alignment step (Stage 3)
* We have chosen to use **STAR** to align reads 
* We have chosen to use **Salmon** to estimate transcript abundance
* We only have access to **2 CPUs and 8Gb** of RAM today 
* We are working with our own subset data today (including reference data)

![](../figs/2.1_RNAseq_experiment.png)

For the sake of expediency, we are using prepared subset data for this session. All the data (including fastqs, input manifest, reference fasta, gtf, and STAR indexes) are available on an external file system called CernVM-FS. CernVM-FS is a read-only file system that Pawsey have used to store files such as containerised tools ([Biocontainers](https://biocontainers.pro/)), reference datasets, and other shared resources that are commonly used by many researchers. Take a look [here](https://support.pawsey.org.au/documentation/display/US/Nimbus+for+Bioinformatics) for more information on bioinformatics resources provided by Pawsey on Nimbus. 

Take a quick look at the workshop data we're working with today: 

```default
ls /path/to/aarnet-cvmfs/training/workshopMaterials
```

We need to store this path in a variable for our run command:
```default
materials=/path/to/aarnet-cvmfs/training/workshopMaterials
```

::: callout-tip
### **Challenge**{.unlisted}

Using the [nf-core/rnaseq documentation](https://nf-co.re/rnaseq/3.11.1/parameters) and the important considerations above, can you decide which flags you may need to add to this command for this experiment?  
```default
nextflow run rnaseq/main.nf \
  --input <samples.tsv> \
  -profile singularity \
  -with-report execution_report_exercise2_1.html \
  -with-trace execution_trace_exercise2_1.txt \
  -with-timeline timeline_exercise2_1.html \
  -with-dag dag_exercise2_1.png
```

:bulb: You will need to look at the [reference genome](https://nf-co.re/rnaseq/3.11.1/parameters#reference-genome-options), [alignment](https://nf-co.re/rnaseq/3.11.1/parameters#alignment-options), and [max job request](https://nf-co.re/rnaseq/3.11.1/parameters#max-job-request-options) sections. 
:::

::: {.callout-caution collapse="true"}
### Solution

Given we are using STAR and Salmon as our aligner and quantification tool of choice (respectively) and it is the default choice of this workflow we will not need to provide an `--aligner` flag. However, if you wanted to provide this for the sake of reproducibility in case things change in the future: 
```default
--aligner 'star_salmon'
```

Given we are providing our own subset data for this workshop, we will need to use:
```default
--fasta /path/to/mouse.fa  
--gtf /path/to/mouse.gtf 
--star_index /path/to/STAR
```

Given we have limited computing resources today, we will need to specify a ceiling for both memory and CPUs:
```default
--max_memory '6.GB' 
--max_cpus 2 
```

:::

### **Run the workflow**

Now run the workflow: 

```default
nextflow run rnaseq/main.nf \
  --input $materials/samples.tsv \
  -profile singularity \
  --fasta $materials/mm10_chr18.fa \
  --gtf $materials/mm10_chr18.gtf \
  --star_index $materials/STAR \
  --max_memory '6 GB' --max_cpus 2 \
  --outdir ex1_results \
  -with-report execution_report_exercise2_1.html \
  -with-trace execution_trace_exercise2_1.txt \
  -with-timeline timeline_exercise2_1.html \
  -with-dag dag_exercise2_1.png
```

:::{.callout-important}
## Zoom reaction check in!
**Is everyone ok? Is your workflow running?**   
:clap: (clap) yes, let's move on.  
:cry: (cry) no, please help. 
:::

### **Explore the workflow code**

While the workflow runs (~15 mins), let's take a closer look at the `rnaseq/workflows/rnaseq.nf` file to see where our flags are influencing the workflow structure. Open a new terminal window so we don't disturb our running workflow, navigate back to `~/nfcore-workshop/session2`, then open the `rnaseq/workflows/rnaseq.nf` file. If you're working on the command-line rather than VScode, run: 

```default 
cat rnaseq/workflows/rnaseq.nf
```

This file is huge and has a lot going on. TODO tie this in with Session 1 content. 

Extract the headers:

```default
sed -n '/\/\*/,/\*\//p' rnaseq/workflows/rnaseq.nf
``` 

We've got 7 main section: 

1. Validate inputs
2. Config files 
3. Import local modules/subworkflows 
4. Import nf-core modules/subworkflows 
5. Run main workflow 
6. Completion email and summary 
7. The end 

#### **Validate inputs**

Here, the developers have set some rules around defining valid input parameters and structuring the workflow to accommodate specified parameters before the workflow is run. 

#### **Config files**

This is just some channels for creating the MultiQC reports, nothing important. 

#### **Import local modules/subworkflows**

#### **Import nf-core modules/subworkflows**

#### **Run main workflow**

#### **Completion email and summary**

#### **The end**

<div class="keypoints">

### **Key points**{.unlisted}

- nf-core workflows are provided with sensible defaults. You can adjust some settings as required by applying flags to your run command.
- nf-core workflows are all built from a template that means they have a standard structure to their code bases  

</div>  