---
title: "**Designing your run command**"
output:
  html_document:
    toc: false
    toc_float: false
from: markdown+emoji
---

::: callout-tip
### Objectives{.unlisted}

- Use the nf-core documentation to select appropriate parameters for a run command 
- Write and run a nf-core rnaseq command on the command line 
- Explore workflow deployment and set up 
:::

### **Download the workflow code**

It can be very easy to lose track while working on the command line, especially when we're working with large datasets and complex commands as we do with bioinformatics workflows. To make sure we work reproducibly, we will be organising our workspace and using a local copy of the [nf-core/rnaseq](https://nf-co.re/rnaseq) workflow for all exercises. 

Start by creating a new directory for all of today's activities and move into it: 

```bash
mkdir ~/nfcore-workshop/session2 && cd $_
```

There are a number of ways to download a nf-core workflow to your machine. We recommend using git or the [nf-core tools](https://nf-co.re/tools/) utility. Today, we will download most recent version (3.11.1) of the workflow from it's [GitHub repository](https://github.com/nf-core/rnaseq) with git. 

Clone the nf-core/rnaseq repository:
```bash
git clone https://github.com/nf-core/rnaseq.git
```

Inside your `nf-core-rnaseq` workflow directory, you should see a number of files and subdirectories: 

```default
ls -l rnaseq
```

```default
total 216
-rw-rw-r-- 1 ubuntu ubuntu 58889 Apr  4 03:40 CHANGELOG.md
-rw-rw-r-- 1 ubuntu ubuntu  9681 Apr  4 03:40 CITATIONS.md
-rw-rw-r-- 1 ubuntu ubuntu  9078 Apr  4 03:40 CODE_OF_CONDUCT.md
-rw-rw-r-- 1 ubuntu ubuntu  1096 Apr  4 03:40 LICENSE
-rw-rw-r-- 1 ubuntu ubuntu 10002 Apr  4 03:40 README.md
drwxrwxr-x 3 ubuntu ubuntu  4096 Apr  4 03:40 assets
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 bin
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 conf
drwxrwxr-x 3 ubuntu ubuntu  4096 Apr  4 03:40 docs
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 lib
-rwxrwxr-x 1 ubuntu ubuntu  2736 Apr  4 03:40 main.nf
drwxrwxr-x 4 ubuntu ubuntu  4096 Apr  4 03:40 modules
-rw-rw-r-- 1 ubuntu ubuntu 13970 Apr  4 03:40 modules.json
-rw-rw-r-- 1 ubuntu ubuntu 10903 Apr  4 03:40 nextflow.config
-rw-rw-r-- 1 ubuntu ubuntu 42576 Apr  4 03:40 nextflow_schema.json
-rw-rw-r-- 1 ubuntu ubuntu   359 Apr  4 03:40 pyproject.toml
drwxrwxr-x 4 ubuntu ubuntu  4096 Apr  4 03:40 subworkflows
-rw-rw-r-- 1 ubuntu ubuntu  1684 Apr  4 03:40 tower.yml
drwxrwxr-x 2 ubuntu ubuntu  4096 Apr  4 03:40 workflows
```

This may look very different to other Nextflow workflows you may have run or written before. The most important files and directories for us to understand, are:

|Feature                |Importance                                                         |
|-----------------------|-------------------------------------------------------------------|
|`conf/`                |Contains standard configuration files for various profiles that build on global settings set by `nextflow.config`|
|`main.nf`              |The executable Nextflow script that defines the structure and flow of the workflow. It calls `workflows/rnaseq.nf`|
|`modules/`             |Contains Nextflow processes used by the workflow. They are called by the `main.nf` file|
|`workflows/rnaseq.nf`  |All the modules, subworkflows, channels, workflow structure for running the rnaseq workflow |

::: {.callout-caution collapse="true"}
### **Alternate installation method**{.unlisted}

Using the nf-core tools utility, search for the rnaseq pipeline: 
```bash
nf-core list rnaseq
```

Then, download the correct pipeline: 
```bash
nf-core download nf-core/rnaseq
```

You will be prompted to select a version. Use your arrow keys to specify 3.10.1 and hit enter. 

> :no_entry: **BEWARE** :no_entry: this method will download a copy of the workflow with a different directory name and slightly different structure. If you choose to use this method, you will need to adjust commands in the upcoming lessons accordingly. 
:::

### **Build your run command**

All nf-core workflows are provided with sensible default settings that have broad applicability and comprensive documentation that explains all available parameters. In the case of the nf-core/rnaseq workflow, parameters are grouped based on various stages of the workflow:

1. **Input/output options** for specifying which files to process and where to save results
2. **UMI options** for processing reads with unique molecular identifiers (UMI)
3. **Read filtering options** to be run prior to alignment
4. **Reference genome options** related to pre-processing of the reference FASTA
5. **Read trimming options** prior to alignment
6. **Alignment options** for read mapping and filtering criteria
7. **Process skipping options** for adjusting the processes to run with the workflow

On the command line you can view these options by running:
```default
nextflow run ../rnaseq/main.nf --help 
```

Notice at the bottom of the print out, there is:
```default
!! Hiding 24 params, use --show_hidden_params to show them !!
```

Three additional parameter sections are hidden from view. This is because they are less commonly used. They include:

8. **Institutional config options** for various compute environments
9. **Max job request options** for limiting memory and cpu usage based on what is available to you
10. **Generic options** focused on how the pipeline is run

To view all the workflow run options on the command line, run:
```default
nextflow run ../rnaseq/main.nf --help --show_hidden_params
```

::: {.callout-note}
### **Hyphens matter!**
Hyphens matter when it comes to parameter flags in nf-core workflows! Nextflow command-line parameters use one (-), whereas pipeline-specific parameters use two (--). For example: -profile is a **Nextflow** parameter, while --input is an **nf-core** parameter.
:::

We will be using a number of flags from each parameters options section. In addition to the required parameters, we have chosen to use some additional flags that are suitable for our experiment. What is 'suitable' varies dramatically between different experiments, computing environments, and datasets, so these settings might not meet your needs for other experiments. Consider the most important design choices we have made for this workflow and structuring our run command:

* We **don't** need to run the pseudo alignment step (Stage 3)
* We have chosen to use **STAR** to align reads 
* We have chosen to use **Salmon** to estimate transcript abundance
* We only have access to **2 CPUs and 8Gb** of RAM today 
* We have already **provided the requisite reference data** (fasta, gtf, indexes) for our training dataset 

![](../figs/2.1_RNAseq_experiment.png)

For the sake of expediency, we are using prepared subset data for this session. All the data (including fastqs, input manifest, reference fasta, gtf, and STAR indexes) are available on an external file system called CernVM-FS. CernVM-FS is a read-only file system that Pawsey have used to store files such as containerised tools ([Biocontainers](https://biocontainers.pro/)), reference datasets, and other shared resources that are commonly used by many researchers. Take a look [here](https://support.pawsey.org.au/documentation/display/US/Nimbus+for+Bioinformatics) for more information on bioinformatics resources provided by Pawsey on Nimbus. 

### **Run the workflow**

We need to store the path to our input and reference data in a variable for our run command:
```default
materials=/path/to/aarnet-cvmfs/training/workshopMaterials
```
Now run the workflow: 

```default
nextflow run rnaseq/main.nf \
    --input $materials/samplesheet.csv \
    -profile singularity \
    --fasta $materials/mm10_reference/mm10_chr18.fa \
    --gtf $materials/mm10_reference/mm10_chr18.gtf \
    --star_index $materials/mm10_reference/STAR \
    --max_memory '6 GB' \
    --max_cpus 2 
```

::: callout-tip
### **Challenge**{.unlisted}

Can you use the [nf-core/rnaseq documentation](https://nf-co.re/rnaseq/3.11.1/parameters) and the important considerations above to explain how we designed our run command for this experiment?

:bulb: You will need to look at the [reference genome](https://nf-co.re/rnaseq/3.11.1/parameters#reference-genome-options), [alignment](https://nf-co.re/rnaseq/3.11.1/parameters#alignment-options), and [max job request](https://nf-co.re/rnaseq/3.11.1/parameters#max-job-request-options) sections. 
:::

::: {.callout-caution collapse="true"}
### Solution

Given we are using STAR and Salmon as our aligner and quantification tool of choice (respectively) and it is the default choice of this workflow we will not need to provide an `--aligner` flag. However, if you wanted to provide this for the sake of reproducibility in case things change in the future: 
```default
--aligner 'star_salmon'
```

Given we are providing our own subset data for this workshop, we will need to use:
```default
--fasta /path/to/mouse.fa  
--gtf /path/to/mouse.gtf 
--star_index /path/to/STAR
```

Given we have limited computing resources today, we will need to specify a ceiling for both memory and CPUs:
```default
--max_memory '6.GB' 
--max_cpus 2 
```
:::

### **Examine the outputs**

Once your workflow has completed, you should see this message printed to your terminal:

```default
[nf-core/rnaseq] Pipeline completed successfully with skipped sampl(es)-
[nf-core/rnaseq] Please check MultiQC report: 6/6 samples failed strandedness check.-
Completed at: 11-Apr-2023 01:01:54
Duration    : 17m 4s
CPU hours   : 0.3
Succeeded   : 200
```

The workflow ran successfully, however note the warning about all samples having failed the strandedness check. We'll explore that in the next lesson. In the meantime, list (`ls -la`) the contents of your directory, you'll see a few new directories (and a hidden directory and log file) have been created:
```default
total 416
drwxrwxr-x   7 ubuntu ubuntu   4096 Apr 11 01:35 .
drwxr-x---  12 ubuntu ubuntu   4096 Apr 11 00:47 ..
drwxrwxr-x   4 ubuntu ubuntu   4096 Apr 11 01:01 .nextflow
-rw-rw-r--   1 ubuntu ubuntu 392378 Apr 11 01:01 .nextflow.log
drwxrwxr-x   7 ubuntu ubuntu   4096 Apr 11 01:01 results
drwxrwxr-x  13 ubuntu ubuntu   4096 Apr  4 03:40 rnaseq
drwxrwxr-x 138 ubuntu ubuntu   4096 Apr 11 00:57 work
```

Nextflow has created 2 new output directories, **work** and **results** in the current directory. 

#### The `work` directory

As each job is run, a unique sub-directory is created in the `work` directory. These directories house temporary files and various command logs created by a process. We can find all information regarding this process that we need to troubleshoot a failed process. 

#### The `results` directory

Depending on the nf-core workflow you are working with, all final outputs will be presented in a directory called `results` by default. You can override this default and output to a directory of your own choosing by using the `--outdir` flag. 

#### The `.nextflow` directory

This directory contains a `cache` subdirectory to store cached data such as downloaded files and can be used to speed up subsequent pipeline runs. It also contains a `history` file which contains a record of pipeline executions including run time, the unique run name, and command line arguments used. 

#### The `.nextflow.log` file

This file is created by Nextflow during the execution of a workflow and contains information about all processes and any warnings or errors that occurred during execution. 

::: {.callout-note}
### **Key points**
- nf-core workflows are provided with sensible defaults. You can adjust some settings as required by applying flags to your run command.
- nf-core workflows are all built from a template that means they have a standard structure to their code bases  
:::

