---
title: "**Using a params file**"
output:
  html_document:
    toc: false
    toc_float: false
  from: markdown+emoji
---

::: callout-tip
### Objectives{.unlisted}

- Use the nf-core documentation to select appropriate parameters for a run command 
- Write and run a nf-core rnaseq command on the command line 
- Explore workflow deployment and set up 
:::

Now that the workflow is 

::: callout-tip
### **Challenge**{.unlisted}

:::

::: {.callout-caution collapse="true"}
### Solution

:::




::: callout-tip
### **Challenge**{.unlisted}

:::

::: {.callout-caution collapse="true"}
### Solution

:::

:::{.callout-important}
## Zoom reaction check in!
**Is everyone ok?**   
:clap: (clap) yes, let's move on.  
:cry: (cry) no, please help. 
:::

### **Run the workflow**{.unlisted}

Make a new working directory for this lesson and move into it: 
```default
mkdir ~/nfcore-workshop/session2/exercise1 && cd $_
```

For the sake of expediency, we are using prepared subset data for this session. All the data (including fastqs, input manifest, reference fasta, gtf, and STAR indexes) are available on the CernVM-FS file system. CernVM-FS is a read-only file system that Pawsey have used to store files such as containerised tools ([Biocontainers](https://biocontainers.pro/)), reference datasets, and other shared resources that are commonly used by many researchers. Take a look [here](https://support.pawsey.org.au/documentation/display/US/Nimbus+for+Bioinformatics) for more information on bioinformatics resources provided by Pawsey on Nimbus. 

Take a quick look at the workshop data we're working with today: 

```default
ls /path/to/aarnet-cvmfs/training/workshopMaterials
```

We need to store this path in a variable for our run command:
```default
materials=/path/to/aarnet-cvmfs/training/workshopMaterials
```

Now run the workflow: 

```default
nextflow run nf-core-rnaseq-3.10.1/workflow/main.nf \
  --input $materials/samples.tsv \
  -profile singularity \
  --fasta $materials/mm10_chr18.fa \
  --gtf $materials/mm10_chr18.gtf \
  --star_index $materials/STAR \
  --max_memory '6 GB' --max_cpus 2 \
  --outdir ex1_results \
  -with-report execution_report_exercise2_1.html \
  -with-trace execution_trace_exercise2_1.txt \
  -with-timeline timeline_exercise2_1.html \
  -with-dag dag_exercise2_1.png
```

### **Reproducibility is a state of mind**{.unlisted}

We have to wait for the workflow to run (this should take ~17 mins) before we can get on with the other exercises. While we wait, lets discuss how we manage reproducibility in our own practices, and share some useful resources with one another:  

![](../figs/Ex2.1_whiteboarding.png)

<div class="keypoints">

### **Key points**{.unlisted}

* 

</div>  



