---
title: "**Specifying external arguments to a process**"
output:
  html_document:
    toc: false
    toc_float: false
from: markdown+emoji
---

::: callout-tip
### Objectives{.unlisted}

- Understand how to use the `ext.args` feature to pass additional command-line arguments to a process
- Implement additional external arguments to a process that are not hardcoded in the process script 
- Observe the behaviour of Nextflow's cache functionality   

:::

nf-core provides a number of flexible parameters for each process in the workflow. These are specified with the double dash syntax (--) on the command line, or saved in a parameters file as we have covered in the previous exercises. However, nf-core has not included every possible parameter that is available to every tool within the workflow; they have included only those most commonly changed from the default as chosen by the software developers.

For example, look at the parameter options for the tool Trim Galore, under [Read trimming options](https://nf-co.re/rnaseq/3.11.1/parameters). You will see 8 parameters, 6 of which are Trim Galore-specific parameters and the last 2 which are nf-core parameters. It would be very unwieldy for nf-core to cover all of the available options for this tool and every other in the workflow on their main workflow page! So, let's look at the full list of trim_galore parameters, either using the [Trim Galore parameters online](https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md#general-options) or using the command line help: 

```default
singularity run work/singularity/depot.galaxyproject.org-singularity-trim-galore-0.6.7--hdfd78af_0.img trim_galore --help | more
```

::: callout-warning
## Exit the Singularity-wrapped TrimGalore command 
To exit the singularity-wrapped trimgalore help command git `Esc` on your keyboard. 
:::

You have the flexibility to use ANY tool parameter, whether a parameter is specified by the nf-core workflow parameters page or not, by providing these as extra arguments.

View the Trim Galore process `main.nf` file:
```default
cat -n rnaseq/modules/nf-core/trimgalore/main.nf
```

Look at the top of the `script:` block (lines 24-25):
```default
script:
def args = task.ext.args ?: ''
```

We can use the ext.args variable to parse any tool parameters to the tool command run by the process. Neat huh! Now look at the actual trim_galore run command, starting at line 42 for single-end reads input: 

```default 
trim_galore \\
  ${args_list.join(' ')} \\
  --cores $cores \\
  --gzip \\
  ${prefix}.fastq.gz
```

Or line 58 for paired-end reads input: 

```default
trim_galore \\
  $args \\
  --cores $cores \\
  --paired \\
  --gzip \\
  ${prefix}_1.fastq.gz \\
  ${prefix}_2.fastq.gz
```

We can see that nf-core has hardcoded the parameters `--cores` and `--gzip`, and also fed in the paired argument if the input is detected as paired-end reads.

As a researcher, its up to you to decipher what customisations are required to make your analysis suitable for your data and research questions. In order to understand what parameters you may want to specify as extra arguments, its first important to: 

1. Read the tool documentation to understand all the available parameters for that tool - tool documentation 
2. View the list of nf-core parameters for that tool - nf-core workflow params page 
3. View which paramteres are hard-coded within the nextflow process - process main.nf

Then, if parameters you want to use are not covered by the nf-core parameters or hard-coded in the run command, you can add them using the `ext.args` variable.

For the sake of the exercise, let's assume we want to increase the minimum quality Phred score from the default of 20 to a much more stringent value of 40 (**note for RNAseq researchers, this of course is way too high!**).

Following the Trim Galore user guide, we would do this with the flag syntax:
```default
--quality 40
```

To parse that to the Trim Galore process, we need to specify this in a custom config, just like we have done for MultiQC in the previous exercises. The difference this time is that we want to restrict the usage of the parameter to the only place it applies, the Trim Galore process. We can do this by using the `withName:` syntax. 

Open a new file `extra_args.config` for editing, then add the following content:
```default
process {
    withName: '.*:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE' {   
        ext.args = '--quality 40'
    }
}
```

We need to instruct nextflow to use the custom config, and we do this with the -c flag:

```default 
nextflow run rnaseq/main.nf \
	-c pawsey_nimbus.config,extra_args.config \
	-profile singularity,c2r8 \
	-params-file params.yaml \
	-resume \
	--outdir Exercise5
```

::: callout-tip
### **Challenge**{.unlisted}

Given we have applied the `-resume` flag, what tasks do you expect to be re-run, and what outputs do you expect to be taken from cache?
:::

::: {.callout-caution collapse="true"}
### Solution

* Taken from cache: initial FastQC processes for each fastq file
* Rerun: read trimming and all downstream processes
:::

Open the file `multiqc_report.html` (use Liver Server in VS Code or scp to take a local copy). On the navigation headings on the left, click on Fail trimmed samples. Observe that all 6 samples have < 5,000 reads remaining after trimming. We can also see this information in a plain text output file:

```default
cat Exercise5/multiqc/star_salmon/multiqc_datamultiqc_fail_trimmed_samples-plot.txt 
```

What has caused these samples to fail? Let's take a look at the [workflow execution file](https://github.com/nf-core/rnaseq/blob/master/workflows/rnaseq.nf):

```default
cat -n rnaseq/workflows/rnaseq.nf
```

Observe line 267:
```default
Filter channels to get samples that passed minimum trimmed read count"
```

and line 295:
```default
if (num_reads <= params.min_trimmed_reads)
```

Our samples have less reads after trimming than required by the parameter `min_trimmed_reads` which has a default value of 10,000.

This highlights the need to **thoroughly check your output to ensure that the intended anlysis has been run and the results are what you require**. The message `Pipeline completed successfully` printed to your terminal, everytime a run completes (and also exit status of zero for invidual tasks or cluster jobs if you are running on a cluster) indicates that there were no errors running the pipeline, **not** that your samples have produced the desired output.

Also note that nf-core has not reported the custom extra argument provided to TrimGalore. On your `multiqc_report.html` file, view the section **nf-core/rnaseq Workflow Summary**. The quality parameter is not described. We can view the full parameters that were supplied to trimgalore in two ways: 

1. View one of the Trim Galore log files for a sample

```default
more Exercise5/trimgalore/SRR3473988.fastq.gz_trimming_report.txt
```
* Note that all supplied parameters, including our custom quality threshold of 40, are listed under 'SUMMARISING RUN PARAMETERS'

2. View the process execution script

Define the run_name and tool variables, like we did in Exercise 2:
```default
run_name=<ENTER_YOUR_RUN_NAME>
tool=trim_galore
```

The run the following custom bash command again: 
```default
nextflow log ${run_name} | while read line;
    do
    cmd=$(ls ${line}/.command.sh 2>/dev/null);
      if grep -q $tool $cmd;
      then  
        echo $cmd;     
      fi; 
    done 
```

* Recall that every task that is run is recorded in a file .command.sh within the task execution directory.
* Use the nextflow log tool to find the name of this run (unless you have remembered it!)

Open the file or use `cat` or a text editor to view it.

< insert screenshot instead of code>

```default
#!/bin/bash -ue
[ ! -f  SRR3473988.fastq.gz ] && ln -s SRR3473988_selected.fastq.gz SRR3473988.fastq.gz
trim_galore \
    --quality 40 \
    --cores 1 \
    --gzip \
    SRR3473988.fastq.gz

cat <<-END_VERSIONS > versions.yml
"NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE":
    trimgalore: $(echo $(trim_galore --version 2>&1) | sed 's/^.*version //; s/Last.*$//')
    cutadapt: $(cutadapt --version)
END_VERSIONS
```

If this were a real analysis, we would re-run with a reduced quality threshold of 30, which is higher than the original default of 20 thus ggiving us that additional stringency we were after while hopefully not being too high to fail all of the samples!

::: {.callout-note}
### **Key points**
- takeaway 1
- takeaway 2
:::
  